# CAS BDAI Individual Innovation Project: Tennis Match Predictor

## Table of Contents 
1. [Introduction](#introduction)
2. [Preliminary steps](#preliminary-steps)
3. [ ](# )
4. [ ](# )
5. [ ](# )
6. [ ](# )
7. [ ](# )
8. [ ](# )


## Introduction <a name="introduction"></a>

### Tennis Match Predictor: GAImeSetMatch


### Goal of this project

...

![.png](img/project/image.png)

Source: [something](https://example.com/)

## Preliminary steps <a name="preliminary-steps"></a>

### Import the dependencies
First we need to import the required libraries: pandas, numpy and matplotlib.pyplot.


In [181]:
import pandas as pd
import numpy as np
import datetime as dt
import matplotlib.pyplot as plt
%matplotlib inline

### Set static parameters
Here we set some parameters which won't be changed. This allows for more easy handling and viewing of the data being explored.

In [182]:
# first, set some static parameters and options (used later too for loading other files)

# directory containing the .csv files
dirname = 'data'

# set options for pandas viewing
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.3f' % x)
# pd.reset_option('display.float_format')

### Define some helper functions
These will help us later with common tasks.

#### Sample data 5 observations

In [183]:

# Small sample dataframe (5 matches) for misc usage
sample_matches_df = pd.DataFrame(data = {
    'tourney_id': ['2000-301', '2000-301', '2000-301', '2000-301', '2000-301'],
    'tourney_name': ['Auckland', 'Auckland', 'Auckland', 'Auckland', 'Auckland'],
    'surface': ['Hard', 'Hard', 'Hard', 'Hard', 'Hard'],
    'draw_size': [32, 32, 32, 32, 32],
    'tourney_level': ['A', 'A', 'A', 'A', 'A'],
    'tourney_date': [20000110, 20000110, 20000110, 20000110, 20000110],
    'match_num': [1, 2, 3, 4, 5],
    'winner_id': [103163, 102607, 103252, 103507, 102103],
    'winner_seed': [1.0, None, None, 7.0, None],
    'winner_entry': [None, 'Q', None, None, 'Q'],
    'winner_name': ['Tommy Haas', 'Juan Balcells', 'Alberto Martin', 'Juan Carlos Ferrero', 'Michael Sell'],
    'winner_hand': ['R', 'R', 'R', 'R', 'R'],
    'winner_ht': [188.0, 190.0, 175.0, 183.0, 180.0],
    'winner_ioc': ['GER', 'ESP', 'ESP', 'ESP', 'USA'],
    'winner_age': [21.7, 24.5, 21.3, 19.9, 27.3],
    'loser_id': [101543, 102644, 102238, 103819, 102765],
    'loser_seed': [None, None, None, None, 4.0],
    'loser_entry': [None, None, None, None, None],
    'loser_name': ['Jeff Tarango', 'Franco Squillari', 'Alberto Berasategui', 'Roger Federer', 'Nicolas Escude'],
    'loser_hand': ['L', 'L', 'L', 'L', 'L'],
    'loser_ht': [180.0, 183.0, 173.0, 185.0, 185.0],
    'loser_ioc': ['USA', 'ARG', 'ESP', 'SUI', 'FRA'],
    'loser_age': [31.1, 24.3, 26.5, 18.4, 23.7],
    'score': ['7-5 4-6 7-5', '7-5 7-5', '6-3 6-1', '6-4 6-4', '0-6 7-6(7) 6-1'],
    'best_of': [3, 3, 3, 3, 3],
    'round': ['R32', 'R32', 'R32', 'R32', 'R32'],
    'minutes': [108.0, 85.0, 56.0, 68.0, 115.0],
    'w_ace': [18.0, 5.0, 0.0, 5.0, 1.0],
    'w_df': [4.0, 3.0, 0.0, 1.0, 2.0],
    'w_svpt': [96.0, 76.0, 55.0, 53.0, 98.0],
    'w_1stIn': [49.0, 52.0, 35.0, 28.0, 66.0],
    'w_1stWon': [39.0, 39.0, 25.0, 26.0, 39.0],
    'w_2ndWon': [28.0, 13.0, 12.0, 15.0, 14.0],
    'w_SvGms': [17.0, 12.0, 8.0, 10.0, 13.0],
    'w_bpSaved': [3.0, 5.0, 1.0, 0.0, 6.0],
    'w_bpFaced': [5.0, 6.0, 1.0, 0.0, 8.0],
    'l_ace': [7.0, 10.0, 6.0, 11.0, 8.0],
    'l_df': [8.0, 7.0, 6.0, 2.0, 8.0],
    'l_svpt': [106.0, 74.0, 56.0, 70.0, 92.0],
    'l_1stIn': [55.0, 32.0, 33.0, 43.0, 46.0],
    'l_1stWon': [39.0, 25.0, 20.0, 29.0, 34.0],
    'l_2ndWon': [29.0, 18.0, 7.0, 14.0, 18.0],
    'l_SvGms': [17.0, 12.0, 8.0, 10.0, 12.0],
    'l_bpSaved': [4.0, 3.0, 7.0, 6.0, 5.0],
    'l_bpFaced': [7.0, 6.0, 11.0, 8.0, 9.0],
    'winner_rank': [1612.0, 211.0, 48.0, 768.0, 167.0],
    'winner_rank_points': [63.0, 157.0, 726.0, 616.0, 219.0],
    'loser_rank': [595.0, 723.0, 649.0, 616.0, 873.0],
    'loser_rank_points': [None, 723.0, 649.0, 616.0, 873.0]
}
)

#### Helper function: Hide winner and loser from columns names
Here's a small sample matches dataframe to be used for miscellaneous usage

In [184]:

def hide_winner_loser(input_df):
    ''' Replace columns starting with 'winner_' and 'loser_' with 'player_1_' and 'player_2_' for the required features
    As we want to be able to predict who will be the winner and the loser in each match, we remove the 'winner_' and 'loser_' columns for each match, 
    and instead replace it with player_1_ and player_2 which are the player names in alphabetical order. 
    The features starting with 'w_' and 'l_' are measures recorded during the match and will not be used in the model for predicting the outcome, so we remove these features
    We will add a column at the end of the dataframe,  which will serve as our y variable '''

    # List of required features
    features = ['id', 'seed', 'entry', 'name', 'hand', 'ht', 'ioc', 'age', 'rank', 'rank_points']
    
    # Copy the input DataFrame to a new one
    df = input_df.copy()

    # Add player_1_name and player_2_name columns
    df['player_1_name'] = df.apply(lambda row: min(row['winner_name'], row['loser_name']), axis=1)
    df['player_2_name'] = df.apply(lambda row: max(row['winner_name'], row['loser_name']), axis=1)

    # Transfer the values from 'winner_' and 'loser_' features to 'player_1_' and 'player_2_' features, according to who was the winner & loser
    for feat in features:
        player_1_feature = np.where(df['player_1_name'] == df['winner_name'],
                                    df['winner_' + feat],
                                    df['loser_' + feat]
                                    )
        player_2_feature = np.where(df['player_2_name'] == df['winner_name'],
                                    df['winner_' + feat],
                                    df['loser_' + feat]
                                    )
        df['player_1_' + feat] = player_1_feature
        df['player_2_' + feat] = player_2_feature   

          
    # Add a winner column
    df['winner'] = df.apply(lambda row: 'player_1' if row['winner_name'] == row['player_1_name'] else 'player_2', axis=1)

    # Remove columns starting with 'winner_' and 'loser_' (they have been replaced by player_1_ and player_2_)
    df = df.loc[:, ~df.columns.str.startswith('winner_') & ~df.columns.str.startswith('loser_')]

    # Remove columns starting with 'w_' and 'l_' (not needed for predicting_)
    df = df.loc[:, ~df.columns.str.startswith('w_') & ~df.columns.str.startswith('l_')]

    return df


#### Test the function hide_winner_loser()

In [185]:
output_df = hide_winner_loser(sample_matches_df)
print(output_df.shape)
output_df

(5, 32)


tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0   2000-301     Auckland    Hard         32             A      20000110   
1   2000-301     Auckland    Hard         32             A      20000110   
2   2000-301     Auckland    Hard         32             A      20000110   
3   2000-301     Auckland    Hard         32             A      20000110   
4   2000-301     Auckland    Hard         32             A      20000110   

   match_num           score  best_of round  minutes        player_1_name  \
0          1     7-5 4-6 7-5        3   R32  108.000         Jeff Tarango   
1          2         7-5 7-5        3   R32   85.000     Franco Squillari   
2          3         6-3 6-1        3   R32   56.000  Alberto Berasategui   
3          4         6-4 6-4        3   R32   68.000  Juan Carlos Ferrero   
4          5  0-6 7-6(7) 6-1        3   R32  115.000         Michael Sell   

    player_2_name  player_1_id  player_2_id  player_1_seed  player_2_seed  \
0      Tommy Haas       101543       103163            NaN          1.000   
1   Juan Balcells       102644       102607            NaN            NaN   
2  Alberto Martin       102238       103252            NaN            NaN   
3   Roger Federer       103507       103819          7.000            NaN   
4  Nicolas Escude       102103       102765            NaN          4.000   

  player_1_entry player_2_entry player_1_hand player_2_hand  player_1_ht  \
0           None           None             L             R      180.000   
1           None              Q             L             R      183.000   
2           None           None             L             R      173.000   
3           None           None             R             L      183.000   
4              Q           None             R             L      180.000   

   player_2_ht player_1_ioc player_2_ioc  player_1_age  player_2_age  \
0      188.000          USA          GER        31.100        21.700   
1      190.000          ARG          ESP        24.300        24.500   
2      175.000          ESP          ESP        26.500        21.300   
3      185.000          ESP          SUI        19.900        18.400   
4      185.000          USA          FRA        27.300        23.700   

   player_1_rank  player_2_rank  player_1_rank_points  player_2_rank_points  \
0        595.000       1612.000                   NaN                63.000   
1        723.000        211.000               723.000               157.000   
2        649.000         48.000               649.000               726.000   
3        768.000        616.000               616.000               616.000   
4        167.000        873.000               219.000               873.000   

     winner  
0  player_2  
1  player_2  
2  player_2  
3  player_1  
4  player_1

## Load and explore the data
This section loads the data available in .csv files from the aforementioned source, explores the data and then cleans it for ease of use and data quality.


### Load matches
Data is available in the form of results of ATP matches. For simplicity reasons, focus only on matches since the year 2000*. Each year is stored in one file using naming convention atp_matches_yyyy.csv.

*The reasoning behind this: since the year 2000, there have been factors that have influenced the outcomes of the modern form of the sport. For me, these are:
1. Racquet technology: Since the 1980s, rackets are made mainly out of graphite. Reference: [Link](https://www.pledgesports.org/2019/08/evolution-of-tennis-rackets/)
2. String technology: In the late 1990s, polyester strings were introduced, which revolutionised the sport. Reference: [Link](https://scientificinquirer.com/2021/08/30/string-theory-the-synthetic-revolution-that-changed-tennis-forever/)
3. Surfaces: in 2009, the ATP discontinued use of carpet court use in all its tournaments. Reference: [Link](https://racketsportsworld.com/tennis-not-played-carpet-courts/#When_was_Carpet_Discontinued_from_Use_in_Tennis)

In [186]:
# create a list of matches (since the year 2000 ) files to load
atp_match_files = [f'{dirname}/atp_matches_{year}.csv' for year in range(2000, 2024)]

In [187]:
# create an empty dataframe to store all matches
matches_df = pd.DataFrame()

# loop through the list of match files, read them and append the data to the combined DataFrame
for filen in atp_match_files:
    matches_df = pd.concat([matches_df, pd.read_csv(filen, index_col=None)])


In [188]:
# explore the matches data
matches_df.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0   2000-301     Auckland    Hard         32             A      20000110   
1   2000-301     Auckland    Hard         32             A      20000110   
2   2000-301     Auckland    Hard         32             A      20000110   
3   2000-301     Auckland    Hard         32             A      20000110   
4   2000-301     Auckland    Hard         32             A      20000110   

   match_num  winner_id  winner_seed winner_entry          winner_name  \
0          1     103163        1.000          NaN           Tommy Haas   
1          2     102607          NaN            Q        Juan Balcells   
2          3     103252          NaN          NaN       Alberto Martin   
3          4     103507        7.000          NaN  Juan Carlos Ferrero   
4          5     102103          NaN            Q         Michael Sell   

  winner_hand  winner_ht winner_ioc  winner_age  loser_id  loser_seed  \
0           R    188.000        GER      21.700    101543         NaN   
1           R    190.000        ESP      24.500    102644         NaN   
2           R    175.000        ESP      21.300    102238         NaN   
3           R    183.000        ESP      19.900    103819         NaN   
4           R    180.000        USA      27.300    102765       4.000   

  loser_entry           loser_name loser_hand  loser_ht loser_ioc  loser_age  \
0         NaN         Jeff Tarango          L   180.000       USA     31.100   
1         NaN     Franco Squillari          L   183.000       ARG     24.300   
2         NaN  Alberto Berasategui          R   173.000       ESP     26.500   
3         NaN        Roger Federer          R   185.000       SUI     18.400   
4         NaN       Nicolas Escude          R   185.000       FRA     23.700   

            score  best_of round  minutes  w_ace  w_df  w_svpt  w_1stIn  \
0     7-5 4-6 7-5        3   R32  108.000 18.000 4.000  96.000   49.000   
1         7-5 7-5        3   R32   85.000  5.000 3.000  76.000   52.000   
2         6-3 6-1        3   R32   56.000  0.000 0.000  55.000   35.000   
3         6-4 6-4        3   R32   68.000  5.000 1.000  53.000   28.000   
4  0-6 7-6(7) 6-1        3   R32  115.000  1.000 2.000  98.000   66.000   

   w_1stWon  w_2ndWon  w_SvGms  w_bpSaved  w_bpFaced  l_ace   l_df  l_svpt  \
0    39.000    28.000   17.000      3.000      5.000  7.000  8.000 106.000   
1    39.000    13.000   12.000      5.000      6.000  5.000 10.000  74.000   
2    25.000    12.000    8.000      1.000      1.000  0.000  6.000  56.000   
3    26.000    15.000   10.000      0.000      0.000 11.000  2.000  70.000   
4    39.000    14.000   13.000      6.000     11.000  8.000  8.000  92.000   

   l_1stIn  l_1stWon  l_2ndWon  l_SvGms  l_bpSaved  l_bpFaced  winner_rank  \
0   55.000    39.000    29.000   17.000      4.000      7.000       11.000   
1   32.000    25.000    18.000   12.000      3.000      6.000      211.000   
2   33.000    20.000     7.000    8.000      7.000     11.000       48.000   
3   43.000    29.000    14.000   10.000      6.000      8.000       45.000   
4   46.000    34.000    18.000   12.000      5.000      9.000      167.000   

   winner_rank_points  loser_rank  loser_rank_points  
0            1612.000      63.000            595.000  
1             157.000      49.000            723.000  
2             726.000      59.000            649.000  
3             768.000      61.000            616.000  
4             219.000      34.000            873.000

In [189]:
# get an overview of number of features, instances, empty values and data types 
matches_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 71213 entries, 0 to 2368
Data columns (total 49 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   tourney_id          71213 non-null  object 
 1   tourney_name        71213 non-null  object 
 2   surface             71213 non-null  object 
 3   draw_size           71213 non-null  int64  
 4   tourney_level       71213 non-null  object 
 5   tourney_date        71213 non-null  int64  
 6   match_num           71213 non-null  int64  
 7   winner_id           71213 non-null  int64  
 8   winner_seed         29586 non-null  float64
 9   winner_entry        8944 non-null   object 
 10  winner_name         71213 non-null  object 
 11  winner_hand         71204 non-null  object 
 12  winner_ht           69582 non-null  float64
 13  winner_ioc          71213 non-null  object 
 14  winner_age          71208 non-null  float64
 15  loser_id            71213 non-null  int64  
 16  loser_seed

In [190]:
matches_df.describe()

draw_size  tourney_date  match_num  winner_id  winner_seed  winner_ht  \
count  71213.000     71213.000  71213.000  71213.000    29586.000  69582.000   
mean      55.128  20109104.958     94.713 108736.809        7.376    186.138   
std       40.045     68421.821    130.189  18210.110        6.790      6.811   
min        2.000  20000103.000      1.000 100644.000        1.000    163.000   
25%       32.000  20050509.000     11.000 103498.000        3.000    183.000   
50%       32.000  20110117.000     28.000 104339.000        5.000    185.000   
75%       64.000  20170203.000    169.000 105227.000        9.000    190.000   
max      128.000  20230828.000   1701.000 211468.000       35.000    211.000   

       winner_age   loser_id  loser_seed  loser_ht  loser_age   best_of  \
count   71208.000  71213.000   16330.000 67939.000  71207.000 71213.000   
mean       26.284 108802.466       8.892   185.599     26.390     3.458   
std         3.959  18259.896       7.328     6.769      4.072     0.840   
min        14.900 100644.000       1.000   163.000     14.500     3.000   
25%        23.400 103444.000       4.000   181.000     23.400     3.000   
50%        26.100 104338.000       7.000   185.000     26.200     3.000   
75%        29.000 105385.000      12.000   190.000     29.200     3.000   
max        42.300 212041.000      35.000   211.000     46.000     5.000   

        minutes     w_ace      w_df    w_svpt   w_1stIn  w_1stWon  w_2ndWon  \
count 63277.000 64811.000 64811.000 64811.000 64811.000 64811.000 64811.000   
mean    106.692     6.912     2.651    77.995    47.971    36.300    16.645   
std      41.178     5.534     2.290    29.239    18.972    13.591     6.980   
min       0.000     0.000     0.000     0.000     0.000     0.000     0.000   
25%      77.000     3.000     1.000    56.000    34.000    27.000    12.000   
50%      99.000     6.000     2.000    73.000    45.000    34.000    16.000   
75%     129.000     9.000     4.000    94.000    58.000    43.000    20.000   
max    1146.000   113.000    26.000   491.000   361.000   292.000    82.000   

        w_SvGms  w_bpSaved  w_bpFaced     l_ace      l_df    l_svpt   l_1stIn  \
count 64812.000  64811.000  64811.000 64811.000 64811.000 64811.000 64811.000   
mean     12.519      3.464      5.038     5.115     3.376    81.034    48.558   
std       4.233      3.078      4.035     4.889     2.535    29.214    19.241   
min       0.000      0.000      0.000     0.000     0.000     0.000     0.000   
25%       9.000      1.000      2.000     2.000     2.000    60.000    35.000   
50%      11.000      3.000      4.000     4.000     3.000    76.000    45.000   
75%      15.000      5.000      7.000     7.000     5.000    97.000    59.000   
max      90.000     24.000     30.000   103.000    26.000   489.000   328.000   

       l_1stWon  l_2ndWon   l_SvGms  l_bpSaved  l_bpFaced  winner_rank  \
count 64811.000 64811.000 64812.000  64811.000  64811.000    70666.000   
mean     32.397    14.965    12.312      4.780      8.628       79.609   
std      14.385     7.207     4.234      3.273      4.148      138.950   
min       0.000     0.000     0.000      0.000      0.000        1.000   
25%      22.000    10.000     9.000      2.000      6.000       18.000   
50%      30.000    14.000    11.000      4.000      8.000       45.000   
75%      40.000    19.000    15.000      7.000     11.000       85.000   
max     284.000   101.000    91.000     27.000     38.000     2101.000   

       winner_rank_points  loser_rank  loser_rank_points  
count           70666.000   69793.000          69793.000  
mean             1592.650     117.938            965.143  
std              1997.658     186.050           1112.623  
min                 1.000       1.000              1.000  
25%               573.000      36.000            426.000  
50%               933.000      68.000            703.000  
75%              1715.000     114.000           1095.000  
max             16950.000    

### Exploring the matches data

#### Zeros
Here we check for zeros in the matches dataframe, in order to decide what to do with them.

In [191]:
# check all features for zero's
zero_count_per_feature= matches_df.apply(lambda col: (col == 0).sum())
zero_count_per_feature

tourney_id                0
tourney_name              0
surface                   0
draw_size                 0
tourney_level             0
tourney_date              0
match_num                 0
winner_id                 0
winner_seed               0
winner_entry              0
winner_name               0
winner_hand               0
winner_ht                 0
winner_ioc                0
winner_age                0
loser_id                  0
loser_seed                0
loser_entry               0
loser_name                0
loser_hand                0
loser_ht                  0
loser_ioc                 0
loser_age                 0
score                     0
best_of                   0
round                     0
minutes                  47
w_ace                  2482
w_df                   9355
w_svpt                    5
w_1stIn                   5
w_1stWon                  7
w_2ndWon                 29
w_SvGms                   9
w_bpSaved             10751
w_bpFaced           

In [192]:
# explore the matches with 0 or less minutes
matches_lessthan_0mins = matches_df.loc[matches_df['minutes']<=0]
matches_lessthan_0mins.head()

tourney_id     tourney_name surface  draw_size tourney_level  \
255    2020-580  Australian Open    Hard        128             G   
361   2020-0891             Pune    Hard         32             A   
376   2020-0506     Buenos Aires    Clay         32             A   
454   2020-0407        Rotterdam    Hard         32             A   
1258  2020-0352    Paris Masters    Hard         64             M   

      tourney_date  match_num  winner_id  winner_seed winner_entry  \
255       20200120        188     126774        6.000          NaN   
361       20200203        287     105216        5.000          NaN   
376       20200210        299     105155          NaN           LL   
454       20200210        275     206173          NaN           WC   
1258      20201102        271     105227          NaN          NaN   

             winner_name winner_hand  winner_ht winner_ioc  winner_age  \
255   Stefanos Tsitsipas           R    193.000        GRE      21.400   
361        Yuichi Sugita           R    173.000        JPN      31.300   
376          Pedro Sousa           R    180.000        POR      31.700   
454        Jannik Sinner           R    188.000        ITA      18.400   
1258         Marin Cilic           R    198.000        CRO      32.000   

      loser_id  loser_seed loser_entry             loser_name loser_hand  \
255     104259         NaN         NaN  Philipp Kohlschreiber          R   
361     104678         NaN           Q         Viktor Troicki          R   
376     106043       1.000         NaN      Diego Schwartzman          R   
454     105430         NaN         NaN             Radu Albot          R   
1258    144895         NaN          WC        Corentin Moutet          L   

      loser_ht loser_ioc  loser_age score  best_of round  minutes  w_ace  \
255    178.000       GER     36.200   W/O        5   R64    0.000    NaN   
361    193.000       SRB     33.900   W/O        3   R16    0.000    NaN   
376    170.000       ARG     27.400   W/O        3    SF    0.000    NaN   
454    175.000       MDA     30.200   W/O        3   R32    0.000    NaN   
1258   178.000       FRA     21.500   W/O        3   R32    0.000    NaN   

      w_df  w_svpt  w_1stIn  w_1stWon  w_2ndWon  w_SvGms  w_bpSaved  \
255    NaN     NaN      NaN       NaN       NaN      NaN        NaN   
361    NaN     NaN      NaN       NaN       NaN      NaN        NaN   
376    NaN     NaN      NaN       NaN       NaN      NaN        NaN   
454    NaN     NaN      NaN       NaN       NaN      NaN        NaN   
1258   NaN     NaN      NaN       NaN       NaN      NaN        NaN   

      w_bpFaced  l_ace  l_df  l_svpt  l_1stIn  l_1stWon  l_2ndWon  l_SvGms  \
255         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
361         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
376         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
454         NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   
1258        NaN    NaN   NaN     NaN      NaN       NaN       NaN      NaN   

      l_bpSaved  l_bpFaced  winner_rank  winner_rank_points  loser_rank  \
255         NaN        NaN        6.000            5375.000      79.000   
361         NaN        NaN       86.000             645.000     191.000   
376         NaN        NaN      145.000             373.000      14.000   
454         NaN        NaN       79.000             710.000      50.000   
1258        NaN        NaN       43.000            1280.000      75.000   

      loser_rank_points  
255             700.000  
361             263.000  
376            2325.000  
454             977.000  
1258            838.000

The matches lasting 0 minutes are all W/O ("Walkovers"), meaning that one player did not contest the match due to injury, illness, etc. These instances should not be used for predicting matches, as they don't measure a player's performance. 

#### NaN or empty values
Here we check for NaN or empty values in the matches dataframe, in order to decide what to do with them.

In [193]:
# check all features for empty values
empty_count_per_feature= matches_df.isnull().sum()
empty_count_per_feature

tourney_id                0
tourney_name              0
surface                   0
draw_size                 0
tourney_level             0
tourney_date              0
match_num                 0
winner_id                 0
winner_seed           41627
winner_entry          62269
winner_name               0
winner_hand               9
winner_ht              1631
winner_ioc                0
winner_age                5
loser_id                  0
loser_seed            54883
loser_entry           56629
loser_name                0
loser_hand               42
loser_ht               3274
loser_ioc                 0
loser_age                 6
score                     0
best_of                   0
round                     0
minutes                7936
w_ace                  6402
w_df                   6402
w_svpt                 6402
w_1stIn                6402
w_1stWon               6402
w_2ndWon               6402
w_SvGms                6401
w_bpSaved              6402
w_bpFaced           

So there are 7 features in the matches dataset which have empty values:
1. minutes
2. seed
3. entry
4. hand
5. ht (height)
6. rank
7. rank_points

In [194]:
# explore the matches with empty minutes
matches_empty_minutes = matches_df.loc[matches_df['minutes'].isnull()]
matches_empty_minutes.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
85    2000-308       Munich    Clay         32             A      20000501   
153   2000-314       Gstaad    Clay         32             A      20000710   
198   2000-315      Newport   Grass         32             A      20000710   
223   2000-316       Bastad    Clay         32             A      20000710   
448   2000-329        Tokyo    Hard         56             A      20001009   

     match_num  winner_id  winner_seed winner_entry        winner_name  \
85          24     101962        2.000          NaN  Younes El Aynaoui   
153          6     103018          NaN          NaN         Max Mirnyi   
198         20     103193          NaN          NaN      Peter Wessels   
223         14     103084          NaN            Q    Guillermo Canas   
448          6     102133          NaN          NaN  Laurence Tieleman   

    winner_hand  winner_ht winner_ioc  winner_age  loser_id  loser_seed  \
85            R    193.000        MAR      28.600    101735         NaN   
153           R    196.000        BLR      23.000    103151         NaN   
198           R    196.000        NED      22.100    102184       6.000   
223           R    185.000        ARG      22.600    102720         NaN   
448           R    180.000        ITA      27.900    103792         NaN   

    loser_entry        loser_name loser_hand  loser_ht loser_ioc  loser_age  \
85          NaN  Richard Fromberg          R   196.000       AUS     30.000   
153         NaN  Mariano Zabaleta          R   183.000       ARG     22.300   
198         NaN    David Prinosil          R   185.000       GER     27.300   
223         NaN         Tomas Zib          R   178.000       CZE     24.400   
448          WC        Joji Miyao          U       NaN       JPN     19.300   

                 score  best_of round  minutes  w_ace  w_df  w_svpt  w_1stIn  \
85      6-7(4) 6-3 6-1        3   R16      NaN    NaN   NaN     NaN      NaN   
153  6-7(3) 7-6(3) 6-3        3   R32      NaN    NaN   NaN     NaN      NaN   
198            6-4 7-5        3   R16      NaN    NaN   NaN     NaN      NaN   
223            6-4 6-3        3   R32      NaN    NaN   NaN     NaN      NaN   
448            6-4 7-5        3   R64      NaN    NaN   NaN     NaN      NaN   

     w_1stWon  w_2ndWon  w_SvGms  w_bpSaved  w_bpFaced  l_ace  l_df  l_svpt  \
85        NaN       NaN      NaN        NaN        NaN    NaN   NaN     NaN   
153       NaN       NaN      NaN        NaN        NaN    NaN   NaN     NaN   
198       NaN       NaN      NaN        NaN        NaN    NaN   NaN     NaN   
223       NaN       NaN      NaN        NaN        NaN    NaN   NaN     NaN   
448       NaN       NaN      NaN        NaN        NaN    NaN   NaN     NaN   

     l_1stIn  l_1stWon  l_2ndWon  l_SvGms  l_bpSaved  l_bpFaced  winner_rank  \
85       NaN       NaN       NaN      NaN        NaN        NaN       16.000   
153      NaN       NaN       NaN      NaN        NaN        NaN       58.000   
198      NaN       NaN       NaN      NaN        NaN        NaN      133.000   
223      NaN       NaN       NaN      NaN        NaN        NaN      144.000   
448      NaN       NaN       NaN      NaN        NaN        NaN      142.000   

     winner_rank_points  loser_rank  loser_rank_points  
85             1325.000      61.000            644.000  
153             700.000      43.000            806.000  
198             281.000      67.000            614.000  
223             252.000     119.000            322.000  
448             269.000    1351.000              1.000

In [195]:
# could matches with empty minutes be due to the tourney_level?
print(matches_empty_minutes['tourney_level'].value_counts())

tourney_level
D    5561
A    1427
G     531
M     382
F      35
Name: count, dtype: int64


### Prediction benchmark for matches
In order to evaluate the prediction accuracy of our model, we need a benchmark to compare when predicting the results of matches. One simple way would be to assume that the higher ranked player will always win. This "higher-ranked player win ratio" can easily be calculated using the features available in the original dataset.
We know that some rankings are empty, so we will just substitute a number higher than the max. ranking (which is 2101).

In [200]:
# setting a prediction benchmark, empty ranking means no ranking, so replace with a arbitrary high value
matches_wins_by_ranking_df = matches_df.copy()
matches_wins_by_ranking_df[['winner_rank','loser_rank']] = matches_wins_by_ranking_df[['winner_rank','loser_rank']].fillna(value=10000)

matches_wins_by_ranking_df['winning_player_ranked'] = matches_wins_by_ranking_df.apply(lambda x: "higher" if x['winner_rank'] < x['loser_rank'] else "lower", axis=1)
print(matches_wins_by_ranking_df['winning_player_ranked'].value_counts(normalize=True)*100)


winning_player_ranked
higher   65.638
lower    34.362
Name: proportion, dtype: float64


So we see that for our dataset, the higher ranked player won **65.6%** of all the matches. This will be our benchmark for evaluating the model.

### Load rankings
Data is also available in the form of ranking of ATP players. It may be required to supplement the missing data for current rankings in the matches dataset, for example, a player doesn't have a ranking at the time of playing a match. 

In [111]:
# create a list of rankings (since the year 2000 ) files to load
atp_rankings_files = [f'{dirname}/atp_rankings_{year}.csv' for year in ['00s','10s', '20s', 'current']]

In [112]:
# create an empty dataframe to store all rankings
rankings_df = pd.DataFrame()

# loop through the list of rankings files, read them and append the data to the combined DataFrame
for filen in atp_rankings_files:
    rankings_df = pd.concat([rankings_df, pd.read_csv(filen, index_col=None)])


In [113]:
# explore the rankings data
rankings_df.head()

ranking_date  rank  player  points
0      20000110     1  101736 4135.00
1      20000110     2  102338 2915.00
2      20000110     3  101948 2419.00
3      20000110     4  103017 2184.00
4      20000110     5  102856 2169.00

In [114]:
# get an overview of number of features, instances, empty values and data types 
rankings_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 2140631 entries, 0 to 58510
Data columns (total 4 columns):
 #   Column        Dtype  
---  ------        -----  
 0   ranking_date  int64  
 1   rank          int64  
 2   player        int64  
 3   points        float64
dtypes: float64(1), int64(3)
memory usage: 81.7 MB


In [115]:
# sanity checks on the data (min values, max values, etc.)
rankings_df.describe()

ranking_date       rank     player     points
count    2140631.00 2140631.00 2140631.00 2139882.00
mean    20112972.33     941.10  119768.99     117.06
std        66763.21     547.58   31216.72     455.88
min     20000110.00       1.00  100149.00       1.00
25%     20060213.00     470.00  104128.00       2.00
50%     20110919.00     946.00  105498.00      10.00
75%     20170306.00    1381.00  120568.00      65.00
max     20230911.00    2271.00  212464.00   16950.00

## Data processing and cleaning

### Scope of processing and cleaning
1. Replace the matches' winner and loser columns
2. Matches with 0 or less minutes
3.
4. 

In [65]:
# remove matches with 0 or less minutes
matches_df = matches_df.loc[matches_df['minutes']<0]


### Execute the processing and cleaning

#### Replace the matches' winner and loser columns
Uses the helper function defined earlier

In [ ]:
# replace the winner and loser columns with player_1 and player_2 for the matches dataset
matches_df= hide_winner_loser(matches_df)
matches_df.head()

tourney_id tourney_name surface  draw_size tourney_level  tourney_date  \
0   2000-301     Auckland    Hard         32             A      20000110   
1   2000-301     Auckland    Hard         32             A      20000110   
2   2000-301     Auckland    Hard         32             A      20000110   
3   2000-301     Auckland    Hard         32             A      20000110   
4   2000-301     Auckland    Hard         32             A      20000110   

   match_num           score  best_of round  minutes        player_1_name  \
0          1     7-5 4-6 7-5        3   R32   108.00         Jeff Tarango   
1          2         7-5 7-5        3   R32    85.00     Franco Squillari   
2          3         6-3 6-1        3   R32    56.00  Alberto Berasategui   
3          4         6-4 6-4        3   R32    68.00  Juan Carlos Ferrero   
4          5  0-6 7-6(7) 6-1        3   R32   115.00         Michael Sell   

    player_2_name  player_1_id  player_2_id  player_1_seed  player_2_seed  \
0      Tommy Haas       101543       103163            NaN           1.00   
1   Juan Balcells       102644       102607            NaN            NaN   
2  Alberto Martin       102238       103252            NaN            NaN   
3   Roger Federer       103507       103819           7.00            NaN   
4  Nicolas Escude       102103       102765            NaN           4.00   

  player_1_entry player_2_entry player_1_hand player_2_hand  player_1_ht  \
0            NaN            NaN             L             R       180.00   
1            NaN              Q             L             R       183.00   
2            NaN            NaN             R             R       173.00   
3            NaN            NaN             R             R       183.00   
4              Q            NaN             R             R       180.00   

   player_2_ht player_1_ioc player_2_ioc  player_1_age  player_2_age  \
0       188.00          USA          GER         31.10         21.70   
1       190.00          ARG          ESP         24.30         24.50   
2       175.00          ESP          ESP         26.50         21.30   
3       185.00          ESP          SUI         19.90         18.40   
4       185.00          USA          FRA         27.30         23.70   

   player_1_rank  player_2_rank  player_1_rank_points  player_2_rank_points  \
0          63.00          11.00                595.00               1612.00   
1          49.00         211.00                723.00                157.00   
2          59.00          48.00                649.00                726.00   
3          45.00          61.00                768.00                616.00   
4         167.00          34.00                219.00                873.00   

     winner  
0  player_2  
1  player_2  
2  player_2  
3  player_1  
4  player_1

### Observations
Some initial observations about the x data sets
1. ...
2. ...

## Final Conclusion  <a name="final-concl"></a>

...